File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left=0,right=height.size()-1,max_left=height[0],max_right=height[height.size()-1];
        int ans=0;
        while(left<right) {
            if (max_left<=max_right){
                left+=1;
                ans+=max(0,max_left-height[left]);
                max_left=max(max_left,height[left]);                  
            }else {
                right-=1;   
                ans+=max(0,max_right-height[right]);
                max_right=max(max_right,height[right]);
            }            
        }
        return ans; 
    }
};

File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        max_left=height[0]
        max_right=height[-1]
        l=1
        r=len(height)-2
        ans=0
        while l<=r:
            if max_left<=max_right:
                if max_left>=height[l]:
                    ans+=max_left-height[l]
                else:
                    max_left=height[l]
                l+=1
            else:
                if max_right>=height[r]:
                    ans+=max_right-height[r]
                else:
                    max_right=height[r]
                r-=1
        return ans
            
            